Edit Functionality for the Report Previewer module

m7pr opened the pull request:
---

### Features Included in This PR

**Cards** - added support for
- **removing** (both legacy `ReportCard` and new `ReportDocument`)
- **editing** (`ReportDocument`)
- **creating new cards** (`ReportDocument`)

**Card blocks** within `ReportDocument` cards, allowed
- **editing of text blocks** 
- **deleting blocks**
- **adding new text blocks**

---

### Not Implemented (but worth considering)

- Reordering cards in the previewer.
- Reordering blocks within the `ReportDocument` edit modal.

---

### Points to Discuss

- Since Shiny doesn't support displaying multiple modals simultaneously, editing multiple text blocks can be tedious—should we explore alternatives?
- Should we allow editing of other (non-text) block types?

---

https://github.com/user-attachments/assets/480ef8de-e591-4117-8725-cdf484cf6c2e

averissimo commented:
Looking good @m7pr :100: 

When editing blocks, can you trigger the modal back after finishing editing, i.e. closing editing modal and re-open.

Alternative would be to renderUI the contents of the modal.

Secondary alternative would be to explore if bslib allows for this, maybe @vedhav can provide more insight.

Some UI considerations

- When adding text block it might be nice to say that it supports markdown formatting
- Consider adding a warning when deleting non-text blocks that they cannot be re-added later, only by deleting the card and adding report again
- Consider having an edit button that is disabled for the legacy cards (with a short explanation why it's disabled when hovering it)

donyunardi commented:
Great start, @m7pr 

Couple comments:

### Edit and Delete Buttons
* I think the the button for edit and delete should be on the right and not under the card title.
<img width="500" alt="Screenshot 2025-04-30 at 3 02 50 PM" src="https://github.com/user-attachments/assets/b5970154-a487-4ab5-92bd-a433b27976b6" />

Or does it make sense to have the edit button when the accordion content is shown?
<img width="500" alt="Screenshot 2025-04-30 at 3 12 54 PM" src="https://github.com/user-attachments/assets/5747f9db-9a3c-4b1e-84c5-b223a9c5406c" />

### Add New Card feature
* I am not sure I understand the purpose of the `Adding a new Card` functionality in the Report Previewer tab since user would normally start from one of the teal modules to create the report card. To me, at this point, user should only be allowed to edit the report card.

## Others:
* We should allow user to edit the card's title too.
    <img width="402" alt="Screenshot 2025-04-30 at 3 00 07 PM" src="https://github.com/user-attachments/assets/68fc6425-b19e-46d8-b097-9d96425ae78e" />

* Fix the code section overflow problem:

<img width="500" alt="image" src="https://github.com/user-attachments/assets/9541d7b7-74b3-4a9e-9814-d320e75edb99" />

to:
<img width="500" alt="image" src="https://github.com/user-attachments/assets/7dec9319-0c9f-493b-9648-26124f832539" />

m7pr commented on R/Editor.R:
```diff
@@ -0,0 +1,69 @@
+    new_content <- lapply(seq_along(x), function(i) editor_srv(i, x[[i]]))
+
+    reactive({
+      # todo: it needs to return report_document, not just list
```
```suggestion
```

m7pr commented:
@donyunardi 

- We removed `Add new Card` button.
- Fixed the width of the code chunk.
- Fixed reactivity.

What we can still consider
- moving edit/delete buttons somewhere else
- allow to edit card title

donyunardi commented:
> What we can still consider
> 
> moving edit/delete buttons somewhere else
> allow to edit card title

I will add these to the next increment.

kumamiao commented:
Thanks @m7pr for adding back the comment box when adding cards.

Agree.
> I think the the button for edit and delete should be on the right and not under the card title.

> We should allow user to edit the card's title too.

Minor comment:
Once user click on the Save button, could we close the modal by default?

Question:
In @donyunardi 's comment screenshot below, it looks like we're able to edit/delete each section for a specific card, but it does not seem to be this way when I tested. Am I pulling from the right branches? This feature would be very helpful for users to customize different sections.
`edit_onclick@redesign@main` from `teal.reporter`
`report_redesign_poc@main` from `TMG` 
`redesign@main` from `teal`

![image](https://github.com/user-attachments/assets/abe8f3aa-7d99-4f77-a2b2-5c9d73310244)

m7pr commented:
@kumamiao this branch (in teal.reporter) has been updated on Friday by @gogonzo .
The code was cleaned, but some functionality was for now removed:
- deletion of blocks in edit card modal
- adding new text block in edit card modal

We also have below in TODO https://github.com/insightsengineering/nestdevs-tasks/issues/106
- sorting blocks
- fix sorting of cards

m7pr commented:
There is also one known issue:

If you try to add a card with the same content that was added just before, this card will not be added because it's identical to the previous one.
